local_server.c

```diff
@@ -182,11 +182,11 @@ void* bulk_input() {
     char dest[5];
     // read destination and message from input file
     while (1) {
-        gets(dest);
+        fgets(dest,10,stdin);
         int dest_id = atoi(dest) - 1;
         if (dest_id < 0) return;
         char message[512];
-        gets(&message[sizeof(int)*3]);
+        fgets(&message[sizeof(int)*3],515,stdin);
```

Comment:
maximum size of message is 512 B, with 12 bytes critical info, you should not read more than 500 B into message body. also take care to convert `\n` to `\0`, you should also `man fgets`